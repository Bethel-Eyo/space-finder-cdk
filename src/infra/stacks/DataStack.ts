import { CfnOutput, Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, Table as DynamodbTable, ITable } from "aws-cdk-lib/aws-dynamodb";
import { Bucket, HttpMethods, IBucket, ObjectOwnership } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { getSuffixFromStack } from "../Utils";

export class DataStack extends Stack {
  public readonly spacesTable: ITable;
  public readonly photosBucket: IBucket;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    this.spacesTable = new DynamodbTable(this, "SpacesTable", {
      partitionKey: {
        name: "id",
        type: AttributeType.STRING,
      },
      tableName: `SpacesTable-${suffix}`, // this should be dynamic to avoid tables using same names
    });

    this.photosBucket = new Bucket(this, 'SpaceFinderPhotos', {
      bucketName: `space-finder-photos-${suffix}`,
      cors: [{
        allowedMethods: [
          HttpMethods.HEAD,
          HttpMethods.GET,
          HttpMethods.PUT
        ],
        allowedOrigins: ["*"],
        allowedHeaders: ["*"]
      }],
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
      // accessControl: BucketAccessControl.PUBLIC_READ // said to currently not be working
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      }
    });

    new CfnOutput(this, "SpaceFinderPhotosBucketName", {
      value: this.photosBucket.bucketName
    })
  }
}
