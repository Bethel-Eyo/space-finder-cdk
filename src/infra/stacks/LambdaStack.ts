import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { LambdaIntegration } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Runtime, Tracing } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import path from "path";

interface LambdaStackProps extends StackProps {
  spacesTable: ITable;
}

export class LambdaStack extends Stack {
  public readonly spacesLambdaIntegration: LambdaIntegration;

  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    const spacesLambda = new NodejsFunction(this, "SpacesLambda", {
      runtime: Runtime.NODEJS_20_X,
      handler: "handler",
      entry: path.join(__dirname, "..", "..", "services", "spaces", "handler.ts"),
      environment: {
        TABLE_NAME: props.spacesTable.tableName,
      },
      tracing: Tracing.ACTIVE, // to use this your handler must have made an sdk call in this case - dynamodbClient
      timeout: Duration.minutes(1),
    });

    spacesLambda.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        resources: [props.spacesTable.tableArn],
        actions: [
          "dynamodb:PutItem",
          "dynamodb:Scan",
          "dynamodb:GetItem",
          "dynamodb:UpdateItem",
          "dynamodb:DeleteItem",
        ],
      })
    );

    this.spacesLambdaIntegration = new LambdaIntegration(spacesLambda);
  }
}
