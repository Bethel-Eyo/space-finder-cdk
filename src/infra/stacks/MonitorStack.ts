import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Alarm, Metric, Unit } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Topic } from "aws-cdk-lib/aws-sns";
import { LambdaSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { Construct } from "constructs";
import { join } from "path";
import { getConfig } from "../Config";

const config = getConfig();
export class MonitorStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const webhookLambda = new NodejsFunction(this, "WebhookLambda", {
      runtime: Runtime.NODEJS_20_X,
      handler: "handler",
      entry: join(__dirname, "..", "..", "services", "monitor", "handler.ts"),
      environment: {
        META_WEBHOOKURL: config.META_WEBHOOKURL,
        META_TOKEN: config.META_TOKEN
      }
    });

    const alarmTopic = new Topic(this, "SpacesAlarmTopic", {
      displayName: "SpacesAlarmTopic",
      topicName: "SpacesAlarmTopic",
    });

    alarmTopic.addSubscription(new LambdaSubscription(webhookLambda));

    const spacesApi4xxAlarm = new Alarm(this, "SpaceApi4xxAlarm", {
      metric: new Metric({
        metricName: "4XXError",
        namespace: "AWS/ApiGateway",
        period: Duration.minutes(1),
        statistic: "Sum",
        unit: Unit.COUNT,
        dimensionsMap: {
          ApiName: "SpacesApi",
        },
      }),
      evaluationPeriods: 1,
      threshold: 5,
      alarmName: "SpaceApi4xxAlarm",
    });

    const topicAction = new SnsAction(alarmTopic);
    spacesApi4xxAlarm.addAlarmAction(topicAction);
    spacesApi4xxAlarm.addOkAction(topicAction);
  }
}
