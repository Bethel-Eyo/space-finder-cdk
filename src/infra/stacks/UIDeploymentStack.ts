import { CfnOutput, Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { getSuffixFromStack } from "../Utils";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { join } from "path";
import { existsSync } from "fs";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { Distribution, OriginAccessIdentity } from "aws-cdk-lib/aws-cloudfront";
import { S3Origin } from "aws-cdk-lib/aws-cloudfront-origins";

export class UIDeploymentStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id);

    const suffix = getSuffixFromStack(this);

    const deploymentBucket = new Bucket(this, "UIDeploymentBucket", {
      bucketName: `space-finder-frontend-${suffix}`,
    });

    const uiDir = join(__dirname, "..", "..", "..", "..", "space-finder-frontend", "dist");
    if (!existsSync(uiDir)) {
      console.warn("UI dir not found: " + uiDir);
      return;
    }

    new BucketDeployment(this, "SpacesFinderDeployment", {
      destinationBucket: deploymentBucket,
      sources: [Source.asset(uiDir)],
    });

    /**Ensure that only cloudfront can access content in this bucket*/
    const originIdentity = new OriginAccessIdentity(this, "OriginAccessIdentity");
    deploymentBucket.grantRead(originIdentity);

    const distribution = new Distribution(this, "SpacesFinderDistribution", {
      defaultRootObject: "index.html",
      defaultBehavior: {
        origin: new S3Origin(deploymentBucket, {
          originAccessIdentity: originIdentity,
        }),
      },
    });

    new CfnOutput(this, "SpaceFinderUrl", {
      value: distribution.distributionDomainName,
    });
  }
}
