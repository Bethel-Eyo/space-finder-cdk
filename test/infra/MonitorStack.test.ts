import { App } from "aws-cdk-lib";
import { MonitorStack } from "../../src/infra/stacks/MonitorStack";
import { Match, Template } from "aws-cdk-lib/assertions";

describe("MonitorStack test suite", () => {
  /** The reason why we use the beforeAll as opposed to the regular beforeEach
   * is because the process of synthesizing the stack to become a cloudformation
   * template might take time and we don't want it to be doing that before each
   * test. we don't want our test running time to be very long.
   */
  let monitorStackTemplate: Template;

  beforeAll(() => {
    const testApp = new App({
      outdir: "cdk.out",
    });

    const monitorStack = new MonitorStack(testApp, "MonitorStack");
    monitorStackTemplate = Template.fromStack(monitorStack);
  });

  test("Lambda properties", () => {
    monitorStackTemplate.hasResourceProperties("AWS::Lambda::Function", {
      Handler: "index.handler",
      Runtime: "nodejs20.x",
    });
  });

  test("Sns topic properties", () => {
    monitorStackTemplate.hasResourceProperties("AWS::SNS::Topic", {
      DisplayName: "SpacesAlarmTopic",
      TopicName: "SpacesAlarmTopic",
    });
  });

  test("Sns subscription properties - with matchers", () => {
    monitorStackTemplate.hasResourceProperties(
      "AWS::SNS::Subscription",
      Match.objectEquals({
        Protocol: "lambda",
        TopicArn: {
          Ref: Match.stringLikeRegexp("SpacesAlarmTopic"),
        },
        Endpoint: {
          "Fn::GetAtt": [Match.stringLikeRegexp("WebhookLambda"), "Arn"],
        },
      })
    );
  });

  test("Sns subscription properties - with exact values", () => {
    const snsTopic = monitorStackTemplate.findResources("AWS::SNS::Topic");
    const snsTopicName = Object.keys(snsTopic)[0];

    const lambda = monitorStackTemplate.findResources("AWS::Lambda::Function");
    const lambdaName = Object.keys(lambda)[0];

    monitorStackTemplate.hasResourceProperties("AWS::SNS::Subscription", {
      Protocol: "lambda",
      TopicArn: {
        Ref: snsTopicName,
      },
      Endpoint: {
        "Fn::GetAtt": [lambdaName, "Arn"],
      },
    });
  });
});
