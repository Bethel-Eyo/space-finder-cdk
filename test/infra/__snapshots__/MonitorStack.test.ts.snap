// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitorStack test suite Monitor stack Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SpaceApi4xxAlarm8C38E49E": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "SpacesAlarmTopicAAAA0BE3",
          },
        ],
        "AlarmName": "SpaceApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "SpacesAlarmTopicAAAA0BE3",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "SpacesAlarmTopicAAAA0BE3": {
      "Properties": {
        "DisplayName": "SpacesAlarmTopic",
        "TopicName": "SpacesAlarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "WebhookLambda118EFE5D": {
      "DependsOn": [
        "WebhookLambdaServiceRoleDB3A5765",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "971e28efe850d9afa78c6db5a32ccc5ff17338ce8c08ed26d1cfb76717e18d40.zip",
        },
        "Environment": {
          "Variables": {
            "META_TOKEN": "EAAXH3uRcPZAgBOZCjZBlTW7ZAWz8aNFkpnvaI5ZCqRWajET9s72aYMyeXZAMw0qIrUohJjZBHcZAdnDXZBbtDaUUNyc0NYWmbVYP6DeY7bJlrpTOuKZBdvhQCb0m1t0Q59ImSoESKqYm1tVn75g4qyU3X1hnXHeUnYNTcxNkpKl5p41ON5WaF35JxLJzgqluebP9ZC2oZAsizbNFkjY5oIXgqpMZD",
            "META_WEBHOOKURL": "https://graph.facebook.com/v20.0/374697945721078/messages",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WebhookLambdaServiceRoleDB3A5765",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebhookLambdaAllowInvokeMonitorStackSpacesAlarmTopicB3FB9ABC36E4BBD6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "SpacesAlarmTopicAAAA0BE3",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebhookLambdaServiceRoleDB3A5765": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WebhookLambdaSpacesAlarmTopicE9C03989": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "WebhookLambda118EFE5D",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "SpacesAlarmTopicAAAA0BE3",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`MonitorStack test suite lambda stack Snapshot 1`] = `
{
  "WebhookLambda118EFE5D": {
    "DependsOn": [
      "WebhookLambdaServiceRoleDB3A5765",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "971e28efe850d9afa78c6db5a32ccc5ff17338ce8c08ed26d1cfb76717e18d40.zip",
      },
      "Environment": {
        "Variables": {
          "META_TOKEN": "EAAXH3uRcPZAgBOZCjZBlTW7ZAWz8aNFkpnvaI5ZCqRWajET9s72aYMyeXZAMw0qIrUohJjZBHcZAdnDXZBbtDaUUNyc0NYWmbVYP6DeY7bJlrpTOuKZBdvhQCb0m1t0Q59ImSoESKqYm1tVn75g4qyU3X1hnXHeUnYNTcxNkpKl5p41ON5WaF35JxLJzgqluebP9ZC2oZAsizbNFkjY5oIXgqpMZD",
          "META_WEBHOOKURL": "https://graph.facebook.com/v20.0/374697945721078/messages",
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "WebhookLambdaServiceRoleDB3A5765",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`MonitorStack test suite snsTopic stack Snapshot 1`] = `
{
  "SpacesAlarmTopicAAAA0BE3": {
    "Properties": {
      "DisplayName": "SpacesAlarmTopic",
      "TopicName": "SpacesAlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
